### PR TITLE
Fix bug where certain changes aren't detected by "rush install"

### PR DESCRIPTION
**Pull request by pgonzal**

### CODE REVIEW

File: apps/rush-lib/src/utilities/JsonFile.ts [hunk 1]
```diff
@@ -7,29 +7,72 @@ import * as jju from 'jju';
...
       throw new Error(`Error reading "${jsonFilename}":` + os.EOL + `  ${error.message}`);
     }
   }
 
-  public static saveJsonFile(jsonData: any, jsonFilename: string): void {
+  /**
+   * Saves the file to disk.  Returns false if nothing was written due to options.onlyIfChanged.
+   */
+  // tslint:disable-next-line:no-any
+  public static saveJsonFile(jsonData: any, jsonFilename: string, options?: ISaveJsonFileOptions): boolean {
```
Comment (iclanton) [inline apps/rush-lib/src/utilities/JsonFile.ts]: >options?: ISaveJsonFileOptions [](start = 66, length = 30)

Just set the default value to {}

File: apps/rush-lib/src/utilities/JsonFile.ts [hunk 2]
```diff
@@ -7,29 +7,72 @@ import * as jju from 'jju';
...
+        }
+      }
+    }
+
     fsx.writeFileSync(jsonFilename, normalized);
+
+    const existingContent: string = fsx.readFileSync(jsonFilename).toString();
+    if (existingContent !== normalized) {
+      throw new Error('onlyIfChanged logic is broken');
+    }
```
Comment (iclanton) [inline apps/rush-lib/src/utilities/JsonFile.ts]: Remove this.

File: apps/rush-lib/src/utilities/JsonFile.ts [hunk 3]
```diff
@@ -7,29 +7,72 @@ import * as jju from 'jju';
...
+  public static saveJsonFile(jsonData: any, jsonFilename: string, options?: ISaveJsonFileOptions): boolean {
     const stringified: string = JSON.stringify(jsonData, undefined, 2) + '\n';
     const normalized: string = Utilities.getAllReplaced(stringified, '\n', '\r\n');
+
+    if (options && options.onlyIfChanged) {
+      // Has the file changed?
+      if (fsx.existsSync(jsonFilename)) {
+        try {
+          const existingContent: string = fsx.readFileSync(jsonFilename).toString();
+          if (existingContent === normalized) {
```
Comment (iclanton) [inline apps/rush-lib/src/utilities/JsonFile.ts]: >existingContent === normalized [](start = 14, length = 30)

Test this on mac, and see what happens if the files have different encodings or if one of them contains a Unicode character.

**Comment by iclanton** [comment]
:shipit:

File: apps/rush-lib/src/utilities/JsonFile.ts (hunk 1 above)
Comment (pgonzal) [inline apps/rush-lib/src/utilities/JsonFile.ts]: If I do that, then the caller could still do saveJsonFile(x,y,undefined)

---
In reply to: [116346856](https://github.com/Microsoft/web-build-tools/pull/187#discussion_r116346856) [](ancestors = 116346856)

**Pull request merged**